src/main/java/com/esri/samples/map/mobile_map_search_and_route/MobileMapSearchAndRouteSample.java

```diff
@@ -0,0 +1,316 @@
+          Point mapPoint = mapView.screenToLocation(point);
+
+          // perform a reverse geocode at the clicked location
+          ReverseGeocodeParameters reverseGeocodeParameters = new ReverseGeocodeParameters();
+          reverseGeocodeParameters.setMaxResults(1);
+          ListenableFuture<List<GeocodeResult>> results = mobileMapPackage.getLocatorTask().reverseGeocodeAsync(mapPoint,
+              reverseGeocodeParameters);
+          results.addDoneListener(() -> {
+            try {
+              // show a pin graphic and a callout with the geocode's address
+              List<GeocodeResult> geocodes = results.get();
+              if (geocodes.size() > 0) {
+                GeocodeResult geocode = geocodes.get(0);
+                Point location = geocode.getDisplayLocation();
+
+                // get attributes from the result for the callout
+                String address = geocode.getAttributes().get("Match_addr").toString();
+                HashMap<String, Object> attributes = new HashMap<>();
+                attributes.put("title", address.split(",")[0]);
+                attributes.put("detail", address.substring(address.indexOf(", ") + 2));
+
+                // create a marker for the location
+                Graphic marker = new Graphic(geocode.getDisplayLocation(), attributes, pinSymbol);
+                locationsGraphicsOverlay.getGraphics().add(marker);
+
+                // display the callout at the location
+                Callout callout = mapView.getCallout();
+                callout.setTitle(marker.getAttributes().get("title").toString());
+                callout.setDetail(marker.getAttributes().get("detail").toString());
+                callout.setLeaderPosition(Callout.LeaderPosition.BOTTOM);
+                callout.showCalloutAt(location, new Point2D(0, -24), Duration.ZERO);
+
+                // if the map has transportation network datasets, solve for a route between the last two locations
+                if (!mapView.getMap().getTransportationNetworks().isEmpty() && locationsGraphicsOverlay.getGraphics().size() > 1) {
+                  // use the first transportation network dataset in the map to create a route task
+                  TransportationNetworkDataset networkDataset = mapView.getMap().getTransportationNetworks().get(0);
+                  RouteTask routeTask = new RouteTask(networkDataset);
+                  // create route parameters with the last two locations' stops
+                  List<Graphic> locationGraphics = locationsGraphicsOverlay.getGraphics();
+                  List<Stop> stops = locationGraphics.subList(Math.max(locationGraphics.size() - 2, 0), locationGraphics.size())
+                      .stream()
+                      .map(g -> new Stop((Point) g.getGeometry()))
+                      .collect(Collectors.toList());
+                  RouteParameters routeParameters = routeTask.createDefaultParametersAsync().get();
```

Comment:
Wouldn't the workflow we want to teach here be,
```java
ListenableFuture<RouteParameters> future = routeTask.createDefaultParametersAsync();
future.addDoneListener(() -> {
     try {
          RouteParameters routeParameters = future.get();
           // ...
      } catch (Exception ex) {
           // some alert message
       }
 });
```

If creating default parameters actual took awhile it would pause the application.